Frameworks/CoreText/CTLine.mm

```diff
@@ -22,25 +22,89 @@
 #import <CoreText/CTTypesetter.h>
 #import <CoreFoundation/CFString.h>
 
-#include <memory>
+#import <memory>
 #import <algorithm>
 #import <numeric>
 #import <cwchar>
 #import <vector>
 
+static inline CFRange __CFRangeIntersectionRange(CFRange a, CFRange b) {
+    CFIndex start = std::max(a.location, b.location);
+    CFIndex end = std::min(a.location + a.length, b.location + b.length);
+    if (end <= start) {
+        return CFRangeMake(0, 0);
+    }
+    return CFRangeMake(start, end - start);
+}
+
+static void __CFAtributedStringAppendAttributedString(CFMutableAttributedStringRef attString, CFAttributedStringRef appendString) {
+    CFRange replaceRange = { CFAttributedStringGetLength(attString), 0 };
+    CFAttributedStringReplaceString(attString, replaceRange, CFAttributedStringGetString(appendString));
+
+    CFRange enumerationRange = { 0, CFAttributedStringGetLength(appendString) };
+    CFIndex index = enumerationRange.location;
+    CFIndex prevlen = CFAttributedStringGetLength(attString);
+    CFIndex lenDiff = 0;
+    CFRange currentRange;
+
+    while ((index >= enumerationRange.location) && index < (enumerationRange.location + enumerationRange.length)) {
```

Comment:
Definitely just use CFAttributedStringReplaceAttributedString - I am 100% sure it is sufficient. The reason we don't use it for [NSMutableAttributedString appendAttributedString:] is to reduce the number of abstract functions. Before we made that class more properly abstract, the initial implementation of it DID use ReplaceAttributedString.